Source/WTF/wtf/text/CStringView.h

```diff
@@ -0,0 +1,150 @@
+#include <wtf/HashFunctions.h>
+#include <wtf/StdLibExtras.h>
+#include <wtf/text/ASCIILiteral.h>
+#include <wtf/text/SuperFastHash.h>
+
+namespace WTF {
+
+class PrintStream;
+class String;
+
+class CStringView final {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    operator const char*() const LIFETIME_BOUND { return m_charactersWithNullTerminator.data(); }
+
+    static CStringView fromNullTerminatedStringUnsafe(const char* string LIFETIME_BOUND)
+    {
+        if (!string)
+            return CStringView();
+        return CStringView { unsafeMakeSpan(string, std::char_traits<char>::length(string) + 1) };
+    }
+
+    static CStringView fromNullTerminatedSpan(std::span<const char> spanWithNullTerminator LIFETIME_BOUND)
+    {
+        RELEASE_ASSERT(spanWithNullTerminator.size() && spanWithNullTerminator[spanWithNullTerminator.size() - 1] == '\0');
+        return CStringView { spanWithNullTerminator };
+    }
+
+    WTF_EXPORT_PRIVATE void dump(PrintStream& out) const;
+
+    CStringView() = default;
+    constexpr CStringView(std::nullptr_t)
+        : CStringView()
+    { }
+    CStringView(ASCIILiteral literal LIFETIME_BOUND)
+    {
+        if (!literal.length())
+            return;
+        m_charactersWithNullTerminator = literal.spanIncludingNullTerminator();
+    }
+
+    unsigned hash() const;
+    bool isNull() const { return m_charactersWithNullTerminator.empty(); }
+
+    const char* characters() const LIFETIME_BOUND { return m_charactersWithNullTerminator.data(); }
+    size_t length() const { return !m_charactersWithNullTerminator.empty() ? m_charactersWithNullTerminator.size() - 1 : 0; }
+    std::span<const char> span() const LIFETIME_BOUND { return m_charactersWithNullTerminator.first(length()); }
+    std::span<const LChar> span8() const LIFETIME_BOUND { return byteCast<LChar>(m_charactersWithNullTerminator.first(length())); }
+    std::span<const char> spanIncludingNullTerminator() const LIFETIME_BOUND { return m_charactersWithNullTerminator; }
+    size_t isEmpty() const { return m_charactersWithNullTerminator.size() <= 1; }
+    WTF_EXPORT_PRIVATE String toString() const;
+
+    char operator[](size_t index) const { return m_charactersWithNullTerminator[index]; }
+    char characterAt(size_t index) const { return m_charactersWithNullTerminator[index]; }
+
+    explicit operator bool() const { return !isEmpty(); }
+    bool operator!() const { return isEmpty(); }
+
+    static CStringView deletedValue();
+    bool isDeletedValue() const { return characters() == reinterpret_cast<char*>(-1); }
+
+private:
+    explicit CStringView(std::span<const char> spanWithNullTerminator LIFETIME_BOUND)
+        : m_charactersWithNullTerminator(spanWithNullTerminator)
+    {
+    }
+
+    std::span<const char> m_charactersWithNullTerminator;
+};
+
+inline bool operator==(CStringView a, CStringView b)
+{
+    if (!a || !b)
+        return a.characters() == b.characters();
+    return equalSpans(a.span(), b.span());
```

Comment:
What happens if we're comparing two views with distinct encodings but identical unicode code points?